### PR TITLE
Feature/category admin api

### LINKED ISSUE #186: Connect the Create new Category form with the DataBase
Create the API to save a new category from the form

### PR DESCRIPTION
**Pull request by fdsosa**
Implemented functions in categories API:
- restoreCategory()
- removeCategory()
- editCategory()
- createCategory()
- getCategoryByTitle()
- checkIfTitleExists()
- category_checkIncomingData()

### CODE REVIEW

File: admin/templates/components/categories.php [hunk 1]
```diff
@@ -39,49 +44,71 @@
...
       <div class="form-group">
         <label for="category_title">Titulo:</label>
-        <input name="category_title" id="category_title" type="text" class="<?php fieldHasError('category_title', 'error') ?>" value="<?php bind(getPreformData('category_title', '')) ?>">
+        <input name="category_title" id="category_title" type="text" class="<?php fieldHasError('category_title', 'error') ?>" value="<?php getGlobal('action') === 'edit' ? bind(oneOf(getPreformData('category_title', ''), $current_category->title)) : bind(getPreformData('category_title', '')) ?>">
       </div>
       <div class="form-group">
         <label for="category_dscp">Descripción:</label>
-        <input name="category_dscp" id="category_dscp" type="text" class="<?php fieldHasError('category_dscp', 'error') ?>" value="<?php bind(getPreformData('category_dscp', '')) ?>">
+        <textarea name="category_dscp" id="category_dscp" type="text" class="<?php fieldHasError('category_dscp', 'error') ?>" value="<?php getGlobal('action') === 'edit' ? bind(oneOf(getPreformData('category_dscp', ''), $current_category->description)) : bind(getPreformData('category_dscp', '')) ?>"></textarea>
```
Comment (Miguelslo27) [inline admin/templates/components/categories.php]: Let's keep the names as long as we need to be more clear, in this case, change `category_dscp` with `category_description`

File: admin/templates/components/categories.php [hunk 2]
```diff
@@ -39,49 +44,71 @@
...
       </div>
       <div class="form-group">
         <label for="category_dscp_short">Descripción Corta:</label>
-        <input name="category_dscp_short" id="category_dscp_short" type="text" class="<?php fieldHasError('category_dscp_short', 'error') ?>" value="<?php bind(getPreformData('category_dscp_short', '')) ?>">
+        <textarea name="category_dscp_short" id="category_dscp_short" type="text" class="<?php fieldHasError('category_dscp_short', 'error') ?>" value="<?php getGlobal('action') === 'edit' ? bind(oneOf(getPreformData('category_dscp_short', ''), $current_category->brief_description)) : bind(getPreformData('category_dscp_short', '')) ?>"></textarea>
```
Comment (Miguelslo27) [inline admin/templates/components/categories.php]: Same here, `category_dscp_short` with `category_description_short` or even better `category_brief_description` as in the field coming from the data base

File: admin/templates/components/categories.php [hunk 3]
```diff
@@ -39,49 +44,71 @@
...
-        <input name="category_img_url" id="category_img_url" type="text" class="<?php fieldHasError('category_img_url', 'error') ?>" value="<?php bind(getPreformData('category_img_url', '')) ?>">
+        <input name="category_img_url" id="category_img_url" type="text" class="<?php fieldHasError('category_img_url', 'error') ?>" value="<?php getGlobal('action') === 'edit' ? bind(oneOf(getPreformData('category_img_url', ''), $current_category->images_url)) : bind(getPreformData('category_img_url', '')) ?>">
       </div>
       <div class="form-group">
         <label for="category_parent">Categoría Padre:</label>
-        <select name="category_parent" id="category_parent" type="text" value="<?php bind(getPreformData('category_parent', '')) ?>" <?php empty($parentCategories) ? bind('disabled') : '' ?>>
+        <select name="category_parent" id="category_parent" type="text" value="<?php getGlobal('action') === 'edit' ? bind(oneOf(getPreformData('category_parent', ''), $cat_parent_value)) : bind(getPreformData('category_parent', '')) ?>" <?php empty($parentCategories) || (isset($current_category_childrens) && !empty($current_category_childrens)) ? bind('disabled') : '' ?>>
```
Comment (Miguelslo27) [inline admin/templates/components/categories.php]: Change **childrens** with **children**, the correct plural for **child** is just **children** without the **s** at the end 😃

File: admin/templates/components/categories.php [hunk 4]
```diff
@@ -39,49 +44,71 @@
...
           <?php if (!empty($parentCategories)) : ?>
-            <option selected> - </option>
+            <option value="no-category"> - </option>
             <?php foreach ($parentCategories as $category) : ?>
-              <option value="<?php bind($category->title) ?>"><?php bind($category->title) ?></option>
+              <option value="<?php bind($category->title) ?>+<?php bind($category->id) ?>" <?php bind((isset($cat_parent) && $category->id == $cat_parent->id) ? 'selected' : '')?>><?php bind($category->title) ?></option>
             <?php endforeach ?>
           <?php endif ?>
         </select>
+        <?php if (isset($current_category_childrens) && !empty($current_category_childrens)) : ?>
```
Comment (Miguelslo27) [inline admin/templates/components/categories.php]: Same here whith **children_s_**

File: admin/templates/components/categories.php [hunk 5]
```diff
@@ -39,49 +44,71 @@
...
+          <div class="explication-container">
```
Comment (Miguelslo27) [inline admin/templates/components/categories.php]: Replace **Explication** with **Explanation**

File: core/api/category.php [hunk 6]
```diff
@@ -59,3 +64,242 @@ function getCategoryById($cid)
...
 }
+
+function checkIfTitleExists($title)
+{
+  return getDB()->countOf('categories', "`title` = '$title'") > 0;
+}
+
+function restoreCategory()
+{
+  $cid = getRequestData('input');
```
Comment (Miguelslo27) [inline core/api/category.php]: What is the `input` and where it comes from? Just curious

File: core/api/category.php [hunk 7]
```diff
@@ -59,3 +64,242 @@ function getCategoryById($cid)
...
+  $sql = (
+    "UPDATE
+        `categories` 
+      SET
+        `status` = 1
+      WHERE 
+        `id` = $cid"
+  );
+  if(!getDB()->query($sql)) {
+    $status->errors[] = 'La red no se puedo restaurar, intente de nuevo';
```
Comment (Miguelslo27) [inline core/api/category.php]: The text says `La red no se pudo restaurar` but you are restoring a Category

File: core/api/category.php [hunk 8]
```diff
@@ -59,3 +64,242 @@ function getCategoryById($cid)
...
+    return $status;
+  }
+
+  $status->succeeded = true;
+  $status->success   = 'Red restaurada con éxito';
```
Comment (Miguelslo27) [inline core/api/category.php]: Same here with Red and category

File: core/api/category.php [hunk 9]
```diff
@@ -59,3 +64,242 @@ function getCategoryById($cid)
...
+      "UPDATE
+        `categories`
+      SET 
+        `category_id` = 0
+      WHERE
+        `category_id` = $cid"
+    );
+
+    if(!getDB()->query($sql)) {
+      $status->errors[] = 'La red no se puedo eliminar, intente de nuevo';
```
Comment (Miguelslo27) [inline core/api/category.php]: Same here, red and category

File: core/api/category.php [hunk 10]
```diff
@@ -59,3 +64,242 @@ function getCategoryById($cid)
...
+      "UPDATE
+        `articles`
+      SET 
+        `category_id` = 0
+      WHERE
+        `category_id` = $cid"
+    );
+
+    if(!getDB()->query($sql)) {
+      $status->errors[] = 'La red no se puedo eliminar, intente de nuevo';
```
Comment (Miguelslo27) [inline core/api/category.php]: Same here, red and category

File: core/api/category.php [hunk 11]
```diff
@@ -59,3 +64,242 @@ function getCategoryById($cid)
...
+  $sql = (
+    "UPDATE
+        `categories` 
+      SET
+        `status` = 0
+      WHERE 
+        `id` = $cid"
+  );
+  if(!getDB()->query($sql)) {
+    $status->errors[] = 'La red no se puedo eliminar, intente de nuevo';
```
Comment (Miguelslo27) [inline core/api/category.php]: Same here, red => category

File: core/api/category.php [hunk 12]
```diff
@@ -59,3 +64,242 @@ function getCategoryById($cid)
...
+    return $status;
+  }
+
+  $status->succeeded = true;
+  $status->success   = 'Red eliminada con éxito';
```
Comment (Miguelslo27) [inline core/api/category.php]: Same here, red => category

File: core/api/category.php [hunk 13]
```diff
@@ -59,3 +64,242 @@ function getCategoryById($cid)
...
+    $parent_category = explode("+", getPostData('category_parent'));
+    $parent_category = $parent_category[1];
+
+    //IF THE PARENT CATEGORY IS ALSO A CHILD CATEGORY RETURN ERROR
+    if (getCategoryById($parent_category)->category_id !== '0') {
+      $status->succeeded        = false;
+      $status->errors           = ['Error, categoria padre inválida'];
+      return $status;
+    }
+  }
```
Comment (Miguelslo27) [inline core/api/category.php]: We are using the very same logic in `editCategory` and in `createCategory`, so maybe we can extract it to a function, let's check on it

File: core/core.php [hunk 14]
```diff
@@ -85,16 +85,20 @@ function processRequests()
     exit;
   }
 
-  if (getRequestData('action') === ACTION_EDIT_SITE_NETWORKS) {
-    setSession('request_messages', siteNetworksEdition());
+  if (getPostData('action') === ACTION_CREATE_CATEGORY) {
+    setSession('request_messages', createCategory());
   }
```
Comment (Miguelslo27) [inline core/core.php]: What will happen with the edit site networks action after this change?

**Review by Miguelslo27 — COMMENTED** [review]
There is some questions, some changes request and some word spelling fix requests :)

File: core/api/category.php (hunk 6 above)
Comment (fdsosa) [reply core/api/category.php]: The input is the category id, I've already changed the variable name.
It comes from the dataset of the restore button, the app.js file have an event listener on that button and comunicate via a fetch method with the api.php file.

Comment (Miguelslo27) [reply core/api/category.php]: Awesome 💪 thanks!

**Comment by Miguelslo27** [comment]
@fdsosa Everything looks great! I will test a bit in my local 👌

**Review by Miguelslo27 — APPROVED** [review]

**Pull request merged**